papis/isbn.py

```diff
@@ -11,24 +17,256 @@
 #: A list of services supported by :mod:`isbnlib`. Note that not all versions
 #: have support for all of these versions.
 ISBN_SERVICE_NAMES = ("goob", "openl", "wiki")
+USER_AGENT = "Mozilla/5.0"
+
+
+def format_title(title: str | None, subtitle: str | None = None) -> str:
+    return (f"{title or ''} - {subtitle}" if subtitle else title or "").strip(
+        ",.:;-_ "
+    )
+
+
+def strip_isbnlike(isbn_like: str) -> str:
+    return "".join(c for c in isbn_like.upper() if c in "0123456789X")
+
+
+def notisbn_fallback(isbn: str | None) -> bool:
+    if isbn is None:
+        return True
+    # Remove non-required symbols such as hyphens or spaces
+    isbn = strip_isbnlike(isbn)
+    # Check length and position of X (only possible for ISBN-10 at position 10)
+    if len(isbn) not in {10, 13} or isbn.find("X") not in {
+        10 * (len(isbn) == 10) - 1,
+        -1,
+    }:
+        return True
+    # Validate checksum digit for ISBN-10
+    if len(isbn) == 10:
+        s = sum((10 - i) * int(isbn[i].replace("X", "10")) for i in range(10))
+        return s % 11 != 0
+    # Check currently available GS1 prefixes for ISBN-13
+    if isbn[0:3] not in {"978", "979"}:
+        return True
+    # Validate checksum digit for ISBN-13
+    return sum((1 + 2 * (i % 2)) * int(isbn[i]) for i in range(13)) % 10 != 0
+
+
+def json_request(url: str) -> Any | None:
+    r = requests.get(url, headers={"user-agent": USER_AGENT})
+    if r.status_code != 200:
+        return None
+    try:
+        data = r.json()
+    except requests.exceptions.JSONDecodeError:
+        return None
+    return data
+
+
+def googlebooks_isbn_search(query: str) -> str:
+    # This assumes the query to be an ISBN
+    query = strip_isbnlike(query)
+    base_api_link = "https://www.googleapis.com/books/v1/volumes?q=isbn:"
+    data = json_request(requests.utils.requote_uri(base_api_link + query))
```

Comment:
This would look nicer if passed a `params={"isbn": isbn}` sort of thing that just gets forwarded to `requests.get`? So we don't have to do the awkward string concatenation / quoting in here.